checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java

```diff
@@ -326,17 +467,376 @@ public Void visitCompoundAssignment(CompoundAssignmentTree node, AnnotatedTypeMi
      * A class for adding annotations based on tree
      */
     private class UnitsTreeAnnotator extends TreeAnnotator {
+        // cache expressions evaluated in evalMathExpression
+        private final Map<String, Pair<Double, Boolean>> mathExpressions;
+
+        private final TypeMirror stringType;
+        private final TypeMirror mathType;
+        private final TypeMirror strictMathType;
 
         UnitsTreeAnnotator(UnitsAnnotatedTypeFactory atypeFactory) {
             super(atypeFactory);
+
+            mathExpressions = new HashMap<String, Pair<Double, Boolean>>();
+
+            stringType = getElementUtils().getTypeElement(java.lang.String.class.getCanonicalName()).asType();
+            mathType = getElementUtils().getTypeElement(java.lang.Math.class.getCanonicalName()).asType();
+            strictMathType = getElementUtils().getTypeElement(java.lang.StrictMath.class.getCanonicalName()).asType();
+        }
+
+        private boolean isSameUnderlyingType(TypeMirror lht, TypeMirror rht) {
+            // use typeUtils.isSameType instead of TypeMirror.equals as this
+            // will check only the underlying type and ignores declarations on
+            // the type mirror
+            return checker.getTypeUtils().isSameType(lht, rht);
         }
 
         @Override
-        public Void visitBinary(BinaryTree node, AnnotatedTypeMirror type) {
-            AnnotatedTypeMirror lht = getAnnotatedType(node.getLeftOperand());
-            AnnotatedTypeMirror rht = getAnnotatedType(node.getRightOperand());
-            Tree.Kind kind = node.getKind();
+        public Void visitMethodInvocation(MethodInvocationTree node, AnnotatedTypeMirror type) {
```

Comment:
This hard-codes a lot of API specific logic in the ATF. Would it be possible to separate more of this into `UnitsRelations` or some new helper class?